allennlp/semparse/contexts/table_question_knowledge_graph.py

```diff
@@ -61,6 +61,87 @@
         **MONTH_NUMBERS,
         }
 
+
+class ColumnTyper():
+    """
+        Type table columns as <string> | <number> | <date> 
+    """
+    pass
+
+
+
+class QuestionEntityExtractor():
+    """
+    A Barebones implementation similar to https://github.com/crazydonkey200/neural-symbolic-machines/blob/master/table/wtq/preprocess.py
+    for extracting entities from a question given a table. 
+    """
+    def __init__(self, 
+                 stop_words_file : str,
+                 table_entity_text: Dict[str, str],
+                 process_conjunction: bool = False) -> None:
+        self.table_entity_text = table_entity_text
+        with open(stop_words_file, 'r') as f:
+            stop_words_list = json.load(f)
+        self.stop_words = set(stop_words_list)
+        self.process_conjunction = process_conjunction
+
+    def _string_in_table(self, candidate_str) -> bool:
+        for entity, text in self.table_entity_text.items():
+            # entity normalized text is followed after fb:cell.
+            if entity.startswith('fb:cell.'):
+            	normalized_text = entity[8:]
+            	if candidate_str in normalized_text: 
+                    return True
+        return False
+
+
+    def _process_conjunction(self, entity_dat):
+        raise NotImplementedError
+
+    def _expand_entities(self, question, entity_dat ):
+        new_ents = []
+        for ent in entity_dat:
+            # to ensure the same strings are not used over and over
+            if new_ents and ent['token_end'] <= new_ents[-1]['token_end']: continue
+            curr_st = ent['token_start']
+            curr_en = ent['token_end']
+            curr_tok = ent['value']
+
+            while curr_en < len(question): 
+                next_tok = question[curr_en]
+                # use the same normalize function used by TableQuestionKnowledgeGraph for consistency
+                next_tok_normalized = TableQuestionKnowledgeGraph._normalize_string(next_tok) 
+                curr_tok_normalized = TableQuestionKnowledgeGraph._normalize_string(next_tok)
+
+                for s, s1, s2 in itertools.product( [' ', '-', '_', ''], [curr_tok, curr_tok_normalized] , [next_tok, next_tok_normalized]  ):
```

Comment:
What is this line doing?

Reply:
If we've already normalized the strings, do we need all of these possible combinations?  And is this going to join `curr_tok` with `next_tok_normalized`?  Is that necessary?

Reply:
I now see that since the cell values are normalized, it probably makes sense to only match it with [curr_tok_normalized, next_tok_normalized] because the others are likely redundant. Do you also agree? I'll make that change

Reply:
Yes, and the normalization code converts hyphens and spaces to underscores, so you don't need to worry about that substitution.